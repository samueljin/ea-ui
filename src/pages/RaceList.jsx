import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Button, Card, Spinner, InputGroup } from "react-bootstrap";
import { racesActions } from '../redux/actions/races.actions';


function RaceList() {

    const ras = useSelector(state => state.races);
    const dispatch = useDispatch();

    const [sortType, setSortType] = useState("ASC");

    useEffect(() => {
        dispatch(racesActions.getRacesByYearAndRound("2021", "1"));
    }, []);

    const handleSort = (e) => {
        (sortType == "ASC") ? setSortType("DES") : setSortType("ASC")
        dispatch(racesActions.sortByPosition(sortType));
    }

    return (
        <div>
            <div class="row">
                <div class="col-10"> Back </div>
                <InputGroup className="col-2">
                    <Button class="btn btn-info" id="button-addon2" onClick={handleSort}>
                        Sort
                    </Button>
                </InputGroup>
            </div>

            {ras.loading && <Spinner animation="border" variant="primary" />}
            {ras.loadingError && <span className="text-danger">ERROR: {ras.error}</span>}

            {ras.items &&
                <div className="grid">
                    {ras.items.map((em, index) => {
                        return (
                            <Card className="d" style={{ width: "15rem" }}>
                                <Card.Body>
                                    <Card.Title>{em.position}</Card.Title>
                                    <ul class="list-group list-group-flush">
                                        <li class="list-group-item">{em.Driver.givenName + " " + em.Driver.familyName}</li>
                                        <li class="list-group-item">points: {em.points}</li>
                                        <li class="list-group-item">  {em.Driver.nationality}</li>
                                        <li class="list-group-item"> car: {em.Constructor.constructorId}</li>
                                    </ul>
                                    <Card.Link href={em.Driver.url}>about...</Card.Link>
                                </Card.Body>
                            </Card>
                        );
                    })}
                </div>
            }
        </div>
    );
}

export default RaceList;