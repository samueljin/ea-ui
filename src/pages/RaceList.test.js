import renderer from 'react-test-renderer';
import RaceList from './RaceList';
import * as reactRedux from 'react-redux'

describe('test suite', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector')
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch')
 // const useHistoryMock = jest.spyOn(reactRedux, 'useHistory')

  beforeEach(() => {
    useSelectorMock.mockClear()
    useDispatchMock.mockClear()
    //useHistoryMock.mockClear()
  })

  it('click sort button to refresh the page', () => {
    const dummyDispatch = jest.fn()
    useDispatchMock.mockReturnValue(dummyDispatch)

    const dummySelector = jest.fn()
    useSelectorMock.mockReturnValue(dummySelector)

    const render = renderer.create(<RaceList/>);
    let tree = render.toJSON();
    expect(tree).toMatchSnapshot();

    // version incompatible  
    //  render.act(() => {
    //    tree.props.handleSort();
    // });
    // re-rendering
    //  tree = render.toJSON();
    //  expect(tree).toMatchSnapshot();

  })
})


  