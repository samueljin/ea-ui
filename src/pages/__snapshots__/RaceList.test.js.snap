// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test suite click sort button to refresh the page 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-10"
    >
       Back 
    </div>
    <div
      className="col-2 input-group"
    >
      <button
        class="btn btn-info"
        className="btn btn-primary"
        disabled={false}
        id="button-addon2"
        onClick={[Function]}
        type="button"
      >
        Sort
      </button>
    </div>
  </div>
</div>
`;
